cpp/src/arrow/compute/kernels/row_encoder_internal.h

```diff
@@ -255,19 +256,19 @@ void IterateListArray(const ArraySpan& arr, ValidFunc&& valid_func,
     return;
   }
   const offset_type* offsets = arr.GetValues<offset_type>(1);
-  const uint8_t* child_data_buffer;
-  if (arr.child_data[0].buffers[1].data == NULLPTR) {
-    child_data_buffer = &empty_value;
-  } else {
-    child_data_buffer = arr.child_data[0].GetValues<uint8_t>(1);
-  }
 
   const uint8_t* child_validity_buffer = arr.child_data[0].buffers[0].data == NULLPTR ?
     NULLPTR : arr.child_data[0].GetValues<uint8_t>(0);
+  const uint8_t* child_data_buffer_1 = arr.child_data[0].buffers[1].data == NULLPTR ?
+    &empty_value : arr.child_data[0].GetValues<uint8_t>(1);
+  const uint8_t* child_data_buffer_2 = arr.child_data[0].buffers[2].data == NULLPTR ?
+    &empty_value : arr.child_data[0].GetValues<uint8_t>(2);
```

Comment:
Yeah it would be much nicer to retrieve the required buffers in the calling function by calling a not-yet-implemented method in `ExecValue data`. And when integrating into `ExecValue` as some Scalar values do not have buffers (like the `Scalar` base class and `PrimitiveScalar`), which makes the implementation a bit trickier. 

I could create a helper function inside of `row_encoder_internal.h` or elsewhere to get the buffers for `ListType` from `ExecValue` specifically, but feel as if it's unclean and unnecessary.  

I do not have any other alternative ways, so I will leave it as is right now. WDYT?